src/lsm/fuzz.zig

```diff
@@ -0,0 +1,459 @@
+                    groove.put(&account);
+                },
+                .get => |get| {
+                    const model_account = model.get(get.id);
+                    const groove = &env.forest.grooves.accounts;
+                    const lsm_account = groove.get(get.id);
+                    if (model_account == null)
+                        assert(lsm_account == null)
+                    else
+                        assert(model_account.?.equal(lsm_account.?));
+                },
+            }
+        }
+    }
+};
+
+fn random_int_exp(comptime Int: type, random: std.rand.Random, avg_int: Int) Int {
+    return @floatToInt(Int, @trunc(random.floatExp(f64) * @intToFloat(f64, avg_int)));
+}
+
+pub fn fuzz(fuzz_ops: []const FuzzOp) !void {
+    // Init mocked storage.
+    var storage = try Storage.init(
+        allocator,
+        Environment.size_max +
+            // TODO this is totally arbitrary
+            16 * 1024 * 1024 * 1024,
+        Storage.Options{
+            // We don't apply storage faults yet, so this seed doesn't matter.
+            .seed = 0xdeadbeef,
+            .read_latency_min = 0,
+            .read_latency_mean = 0,
+            .write_latency_min = 0,
+            .write_latency_mean = 0,
+            .read_fault_probability = 0,
+            .write_fault_probability = 0,
+        },
+        0,
+        .{
+            .first_offset = 0,
+            .period = 0,
+        },
+    );
+    defer storage.deinit(allocator);
+
+    try Environment.format(&storage);
+    try Environment.run(&storage, fuzz_ops);
+}
+
+pub fn generate_fuzz_ops(random: std.rand.Random) ![]const FuzzOp {
+    const fuzz_op_count = random_int_exp(usize, random, 1E6);
+    const fuzz_ops = try allocator.alloc(FuzzOp, fuzz_op_count);
+
+    // We're not allowed to go more than Environment.cache_entries_max puts without compacting.
+    var puts_since_compact: usize = 0;
+
+    for (fuzz_ops) |*fuzz_op| {
+        fuzz_op.* = if (puts_since_compact >= Environment.cache_entries_max)
+            // We have to compact before doing any other operations.
+            FuzzOp{ .compact = {} }
+        else
+        // Otherwise pick a random FuzzOp.
+        switch (random.enumValue(std.meta.Tag(FuzzOp))) {
+            .compact => FuzzOp{
+                .compact = {},
+            },
+            .put => FuzzOp{
+                .put = Account{
+                    // Use exponentially distributed ids to increase the chances of overwriting previous values.
+                    .id = random_int_exp(u128, random, 10),
+                    // TODO It would be nice to make timestamp random, but we have to ensure that they match the timestamp of previous Accounts with the same id.
+                    .timestamp = 0,
+                    .user_data = 0,
+                    .reserved = [_]u8{0} ** 48,
+                    .ledger = random.int(u32),
+                    .code = random.int(u16),
```

Comment:
`ledger`, `code`, and `user_data` are secondary indices — maybe we should use exponential random numbers here as well.

Reply:
:+1:

Reply:
I'll add ops to query secondary indices too.

Reply:
We don't support that yet, except at the tree level. That is going to be part of implementing range queries.

Reply:
Ok!